File: R/stacked_cp_california.R
# Stacked conformal prediction (California Housing)

library(tidyverse)
library(rsample)
library(ranger)
library(catboost)

source("functions.R")

california <- read_csv("california.csv", show_col_types = FALSE) |>
    mutate(ocean_proximity = as_factor(ocean_proximity)) |>
    rename(y = median_house_value)

seed <- 42

set.seed(seed)

split <- initial_split(california, prop = 0.7)

trn <- training(split)
tst <- testing(split)

# Random Forest

rf <- ranger(y ~ ., data = trn, num.trees = 10^3)
y_hat_tst_rf <- predict(rf, data = tst)$predictions
sqrt(mean((tst$y - y_hat_tst_rf)^2))

# Catboost

trn_pool <- catboost.load_pool(data = trn |> select(-y), label = trn$y) 
cb <- catboost.train(trn_pool, params = list(random_seed = seed, verbose = 100)) # 0 = silent
tst_pool <- catboost.load_pool(data = tst)
y_hat_tst_cb <- catboost.predict(cb, tst_pool)
sqrt(mean((tst$y - y_hat_tst_cb)^2))

pred_tst <- tibble(z1 = y_hat_tst_cb, z2 = y_hat_tst_rf, y = tst$y)

# saveRDS(pred_tst, file = "pred_tst_california.RDS")
# pred_tst <- readRDS(file = "pred_tst_california.RDS")

# stacking the models

begin <- Sys.time()

pred <- tibble()
num_folds <- 10
fold <- sample(1:num_folds, size = nrow(trn), replace = TRUE)
pb <- txtProgressBar(min = 1, max = num_folds, style = 3)
for (k in 1:num_folds) {
    out_of_sample <- trn[fold == k, ]
    in_sample <- trn[fold != k, ]
    # CatBoost
    in_pool <- catboost.load_pool(data = in_sample |> select(-y), label = in_sample$y)
    cb <- catboost.train(in_pool, params = list(random_seed = seed, verbose = 0))
    out_pool <- catboost.load_pool(data = out_of_sample)
    # Random Forest
    rf <- ranger(y ~ ., data = in_sample, num.trees = 10^3)
    # base-models predictions become regressors for the meta-learner
    pred <- rbind(pred, tibble(z1 = catboost.predict(cb, out_pool),
                               z2 = predict(rf, data = out_of_sample)$predictions,
                               y = out_of_sample$y))
    setTxtProgressBar(pb, k)
}
close(pb)

print(Sys.time() - begin) # ~ 6 minutes

# saveRDS(pred, file = "pred_california.RDS")
# pred <- readRDS(file = "pred_california.RDS")

meta_learner <- lm(y ~ z1 + z2, data = pred)

y_hat_tst <- predict(meta_learner, newdata = pred_tst)

sqrt(mean((pred_tst$y - y_hat_tst)^2))

###

Z <- model.matrix(meta_learner)

Z_tst <- model.matrix(y ~ z1 + z2, data = pred_tst)

alpha <- 0.1

cp_interval <- full_cp_interval(Z, pred$y, Z_tst, alpha)

# saveRDS(cp_interval, file = "cp_interval_california.RDS")
# cp_interval <- readRDS(file = "cp_interval_california.RDS")

summary(cp_interval[, 2] - cp_interval[, 1])

mean(cp_interval[, 1] <= pred_tst$y & pred_tst$y <= cp_interval[, 2])

File: Python/stacked_cp_california.py
import numpy as np
import pandas as pd
from catboost import CatBoostRegressor, Pool
from sklearn.ensemble import RandomForestRegressor
import statsmodels.formula.api as smf
from sklearn.model_selection import train_test_split
from tqdm.auto import tqdm
from functions import full_cp_interval

california = (
    pd.read_csv("california.csv")
      .assign(ocean_proximity = lambda df: df["ocean_proximity"].astype("category"))
      .rename(columns = {"median_house_value": "y"})
)

seed = 42
trn, tst = train_test_split(california, test_size = 0.30, random_state = seed)

alpha = 0.10

X_trn, y_trn = trn.drop(columns = "y"), trn["y"]
X_tst, y_tst = tst.drop(columns = "y"), tst["y"]

cat_cols = X_trn.select_dtypes(["category"]).columns.tolist()

# preparation of RF data by encoding categories
X_trn_rf = X_trn.copy()
X_tst_rf = X_tst.copy()
for col in cat_cols:
    X_trn_rf[col] = X_trn_rf[col].cat.codes
    X_tst_rf[col] = X_tst_rf[col].cat.codes

rf_full = RandomForestRegressor(n_estimators = 500, random_state = seed, n_jobs = -1)
rf_full.fit(X_trn_rf, y_trn)

trn_pool = Pool(data = X_trn, label = y_trn, cat_features = [X_trn.columns.get_loc(c) for c in cat_cols])
cb_full = CatBoostRegressor(random_seed = seed, verbose = 0)
cb_full.fit(trn_pool)

y_hat_tst_rf = rf_full.predict(X_tst_rf)
y_hat_tst_cb = cb_full.predict(Pool(data = X_tst, cat_features = [X_trn.columns.get_loc(c) for c in cat_cols]))

pred_tst = pd.DataFrame({"z1": y_hat_tst_cb, "z2": y_hat_tst_rf, "y": y_tst})

num_folds = 10
rng = np.random.default_rng(seed)
fold = rng.integers(0, num_folds, size = X_trn.shape[0])

print("Stacking the models...")

rows = []
for k in tqdm(range(num_folds), total = num_folds, desc = "fold", leave = False):
    out_of_sample = (fold == k)
    in_sample = ~out_of_sample
    X_in = X_trn.loc[in_sample]
    y_in = y_trn.loc[in_sample]
    X_out = X_trn.loc[out_of_sample]
    y_out = y_trn.loc[out_of_sample]

    # encode for RF
    X_in_rf = X_in.copy()
    X_out_rf = X_out.copy()
    for col in cat_cols:
        X_in_rf[col] = X_in_rf[col].cat.codes
        X_out_rf[col] = X_out_rf[col].cat.codes

    rf = RandomForestRegressor(n_estimators = 500, random_state = seed, n_jobs = -1)
    rf.fit(X_in_rf, y_in)
    cb = CatBoostRegressor(random_seed = seed, verbose = 0)
    cb.fit(Pool(data = X_in, label = y_in, cat_features = [X_trn.columns.get_loc(c) for c in cat_cols]))
    z1 = cb.predict(Pool(data = X_out, cat_features = [X_trn.columns.get_loc(c) for c in cat_cols]))
    z2 = rf.predict(X_out_rf)
    rows.append(pd.DataFrame({"z1": z1, "z2": z2, "y": y_out.values}))

pred = pd.concat(rows, ignore_index = True)

meta_learner = smf.ols(formula = "y ~ z1 + z2", data = pred).fit()

y_hat_tst = meta_learner.predict(pred_tst[["z1", "z2"]])

Z = np.column_stack((np.ones(pred.shape[0]), pred[["z1", "z2"]].values))
Z_tst = np.column_stack((np.ones(pred_tst.shape[0]), pred_tst[["z1", "z2"]].values))

print("Computing intervals...")

cp_interval = full_cp_interval(Z, pred["y"].values, Z_tst, alpha)

coverage = np.mean((cp_interval[:, 0] <= pred_tst["y"].values) & (pred_tst["y"].values <= cp_interval[:, 1]))
median_width = np.median(cp_interval[:, 1] - cp_interval[:, 0])

print(f"Coverage: {coverage:.3f}")
print(f"Median width: {median_width:.2f}")
